lesson05/circle_test.go

```diff
@@ -0,0 +1,108 @@
+package main
+
+import (
+	"math"
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAreaC(t *testing.T) {
+
+	tests := []struct {
+		name    string
+		args    Circle
+		want    float64
+		wantErr bool
+	}{
+		{
+			name: "if radius is <= 0 should return error ",
+			args: Circle{
+				Radius: 0.0,
+			},
+			want:    0.0,
+			wantErr: true,
+		},
+		{
+			name: "if radius = 1 should return Pi",
+			args: Circle{
+				Radius: 1.0,
+			},
+			want:    math.Pi,
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := (tt.args).Area()
+
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Area() error = %v, wantErr %v, got = %v", err, tt.wantErr, got)
+				assert.Equal(t, err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("Area() = %v, want = %v", got, tt.want)
+				assert.Equal(t, got, tt.want)
+			}
+		})
+	}
+}
+func TestPerimeterC(t *testing.T) {
+
+	tests := []struct {
+		name    string
+		args    Circle
+		want    float64
+		wantErr bool
+	}{
+		{
+			name: "if radius is <= 0 should return error ",
+			args: Circle{
+				Radius: 0.0,
+			},
+			want:    0.0,
+			wantErr: true,
+		},
+		{
+			name: "if radius = 1 should return 2 * Pi",
+			args: Circle{
+				Radius: 1.0,
+			},
+			want:    math.Pi * 2,
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := (tt.args).Perimeter()
+
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Perimeter() error = %v, wantErr %v, got = %v", err, tt.wantErr, got)
+				assert.Equal(t, err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("Perimeter() = %v, want = %v", got, tt.want)
+				assert.Equal(t, got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCircle_String(t *testing.T) {
```

Comment:
This test will work, but for a clearer code style, you should use the same pattern as you used in `TestAreaC` and `TestPerimeterC` - making a slice of test cases and running them in the for loop. Also note, that here you can also test equality of two strings using `assert.Equal`